src/declarativewebpage.cpp

```diff
@@ -114,50 +115,89 @@ void DeclarativeWebPage::setContainer(DeclarativeWebContainer *container)
 
 int DeclarativeWebPage::tabId() const
 {
-    return m_tabId;
+    return m_tab.tabId();
 }
 
-void DeclarativeWebPage::setTabId(int tabId)
+void DeclarativeWebPage::setTab(const Tab& tab)
 {
-    if (m_tabId != tabId) {
-        m_tabId = tabId;
-        emit tabIdChanged();
-    }
-}
+    Q_ASSERT(m_tab.tabId() == 0);
 
-bool DeclarativeWebPage::domContentLoaded() const
-{
-    return m_domContentLoaded;
+    m_tab = tab;
+    emit tabIdChanged();
+    connect(DBManager::instance(), SIGNAL(tabHistoryAvailable(int, QList<Link>)),
+            this, SLOT(onTabHistoryAvailable(int, QList<Link>)));
+    DBManager::instance()->getTabHistory(m_tab.tabId());
 }
 
-bool DeclarativeWebPage::urlHasChanged() const
+void DeclarativeWebPage::onUrlChanged()
 {
-    return m_urlHasChanged;
+    disconnect(this, SIGNAL(urlChanged()), this, SLOT(onUrlChanged()));
+    m_urlReady = true;
+    restoreHistory();
 }
 
-void DeclarativeWebPage::setUrlHasChanged(bool urlHasChanged)
+void DeclarativeWebPage::onTabHistoryAvailable(const int& tabId, const QList<Link>& links)
 {
-    m_urlHasChanged = urlHasChanged;
+    if (tabId == m_tab.tabId()) {
+        m_restoredTabHistory = links;
+
+        // reverse the list
+        for(int k=0, s=m_restoredTabHistory.size(), max=(s/2); k<max; k++) {
+            m_restoredTabHistory.swap(k,s-(1+k));
+        }
```

Comment:
std::reverse should work here:

```
std::reverse(m_restoredTabHistory.begin(), m_restoredTabHistory.end());
```
At least in Qt5.4 plenty of algorithms are marked as obsolete and respected std algorithm recommended instead http://doc.qt.io/qt-5/qtalgorithms-obsolete.html